tests/migrations/test_autodetector.py

```diff
@@ -2361,6 +2361,57 @@ def test_multiple_bases(self):
         self.assertOperationAttributes(changes, "app", 0, 3, name="D")
         self.assertOperationAttributes(changes, "app", 0, 4, name="E")
 
+    def test_metaclass(self):
+        class AlphaBase(type):
+            pass
+
+        class Alpha(metaclass=AlphaBase):
+            pass
+
+        class ABase(AlphaBase, type(models.Model)):
+            pass
+
+        Failure = ModelState(
+            "app",
+            "Failure",
+            [("a_id", models.AutoField(primary_key=True))],
+            bases=(Alpha, models.Model),
+        )
+
+        Failure1 = ModelState(
+            "app",
+            "Failure",
+            [("a_id", models.AutoField(primary_key=True))],
+            bases=(Alpha, models.Model),
+            metaclass=AlphaBase
+        )
+
+        Success = ModelState(
+            "app",
+            "Success",
+            [("a_id", models.AutoField(primary_key=True))],
+            bases=(Alpha, models.Model),
+            metaclass=ABase
+        )
+
+        # test TypeError raised if correct metaclass isn't provided
+        with self.assertRaises(TypeError):
+            changes = self.get_changes([], [Failure])
+        with self.assertRaises(TypeError):
+            changes = self.get_changes([], [Failure1])
+
+        changes = self.get_changes([], [Success])
+        # Right number/type of migrations?
+        self.assertNumberMigrations(changes, "app", 1)
+        self.assertOperationTypes(changes, "app", 0, ["CreateModel"])
+        self.assertOperationAttributes(changes, "app", 0, 0, name="Success")
+
+        # test deletion of model
+        changes = self.get_changes([Success], [])
+        self.assertNumberMigrations(changes, "app", 1)
+        self.assertOperationTypes(changes, "app", 0, ["DeleteModel"])
+        self.assertOperationAttributes(changes, "app", 0, 0, name="Success")
```

Comment:
Can we add tests for detecting changes in metaclasses?

Reply:
@felixxm good point. I'd be willing to do this, but I'm not entirely sure how (the other test I added in was essentially just copy-paste other code). Could you point me to an example of what this test would look like?

Reply:
You can create two states (before and after) and check detected changed, see `test_alter_model_options()`.